src/guest_memory.rs

```diff
@@ -236,6 +237,68 @@ pub trait GuestMemoryRegion: Bytes<MemoryRegionAddress, E = Error> {
     }
 }
 
+struct GuestMemoryRange<'a, R> {
+    region: &'a R,
+    start: MemoryRegionAddress,
+    len: usize,
+}
+
+struct GuestMemoryRangeIter<'a, M: GuestMemory + ?Sized> {
+    memory: &'a M,
+    addr: GuestAddress,
+    count: usize,
+    total: usize,
+
+    // Stores the length of the current GuestMemoryRange to
+    // check for bugs.
+    _last_length: Cell<usize>,
+}
+
+impl<'a, M: GuestMemory + ?Sized> GuestMemoryRangeIter<'a, M> {
+    pub fn new(memory: &'a M, addr: GuestAddress, count: usize) -> GuestMemoryRangeIter<'a, M> {
+        GuestMemoryRangeIter {
+            memory,
+            addr,
+            count,
+            total: 0,
+            _last_length: Cell::new(0),
+        }
+    }
+
+    fn current_region(&self) -> Result<Option<&'a M::R>> {
+        if self.total == self.count {
+            return Ok(None);
+        }
+        match self.memory.find_region(self.addr) {
+            // A partial read is okay, but if self.total is zero the initial
+            // address was invalid
+            None if self.total == 0 => Err(Error::InvalidGuestAddress(self.addr)),
+            opt => Ok(opt),
+        }
+    }
+
+    pub fn current_range(&self) -> Result<Option<GuestMemoryRange<M::R>>> {
+        let range = self.current_region()?.map(|region| {
+            let start = region.to_region_addr(self.addr).unwrap();
+            let cap = region.len() - start.raw_value();
+            let len = std::cmp::min(cap, (self.count - self.total) as GuestUsize) as usize;
+            self._last_length.set(len);
+            GuestMemoryRange { region, start, len }
+        });
+        Ok(range)
+    }
+
+    pub fn advance(&mut self, count: usize) {
+        assert!(count > 0 && count <= self._last_length.get());
+        self.total += count;
+        self.addr = match self.addr.overflowing_add(count as GuestUsize) {
+            (GuestAddress(0), _) => GuestAddress(0),
+            (result, false) => result,
+            (_, true) => panic!("guest address overflow"),
```

Comment:
Can you add a comment here to explain the need for the panic? I remember we discussed about this during KVM Forum as well. I generally don't expect panics in library code.